pkg/rdkafka/RdKafkaContext.php

```diff
@@ -171,6 +177,12 @@ private function getProducer(): VendorProducer
             if (isset($this->config['log_level'])) {
                 $this->producer->setLogLevel($this->config['log_level']);
             }
+
+            // Once created RdKafkaProducer can store messages internally that need to be delivered before PHP shuts
+            // down. Otherwise, we are bound to lose messages in transit.
+            // Note that it is generally preferable to call "close" method explicitly before shutdown starts, since
+            // otherwise we might not have access to some objects, like database connections.
+            register_shutdown_function([$this, 'close']);
```

Comment:
looks like this will be called as many times as `getProduce()` is executed. 

I would suggest to add an early return or something here to call shutdown function only once

Reply:
> looks like this will be called as many times as `getProduce()` is executed.

This is kept inside an `if` statement that only initializes if `$this->producer` has not yet been set, and it's the only place where this property is set. So it will called at most once for each context.

Reply:
I see, didn't notice parent `if`. Thank you